trin-bridge/src/bridge.rs

```diff
@@ -118,118 +114,126 @@ impl Bridge {
+                    }
                 }
+            } else {
+                None
             };
-            let mut gossip_batch = GossipBatch::new(
-                block_range_to_gossip.clone(),
-                Some(epoch_index),
-                full_headers,
-            );
-            if let Err(msg) = self.gossip_headers(&mut gossip_batch).await {
-                warn!("Error gossiping headers in range: {block_range_to_gossip:?} - {msg:?}. Skipping iteration.");
-                epoch_index += 1;
-                continue;
-            };
-            // Sleep for 5 seconds to allow headers to saturate network,
-            // since they must be available for body / receipt validation
-            sleep(Duration::from_secs(HEADER_SATURATION_DELAY)).await;
-            if let Err(err) = self.serve_bodies(&mut gossip_batch).await {
-                warn!("error gossiping bodies: {err:?}");
-            };
-            if let Err(err) = self.serve_receipts(&mut gossip_batch).await {
-                warn!("error gossiping receipts: {err:?}");
-            };
-            gossip_batch.display_stats();
+            info!("fetching headers in range: {gossip_range:?}");
+            let gossip_stats = Arc::new(Mutex::new(GossipStats::new(gossip_range.clone())));
+            let futures = futures::stream::iter(gossip_range.into_iter().map(|height| {
+                let epoch_acc = epoch_acc.clone();
+                let gossip_stats = gossip_stats.clone();
+                async move {
+                    let _ = Bridge::serve_full_block(
+                        height,
+                        epoch_acc,
+                        gossip_stats,
+                        self.portal_clients.clone(),
+                    )
+                    .await;
+                }
+            }))
+            .buffer_unordered(FUTURES_BUFFER_SIZE)
+            .collect::<Vec<()>>();
+            futures.await;
+            if let Ok(gossip_stats) = gossip_stats.lock() {
+                gossip_stats.display_stats();
+            } else {
+                warn!("Error displaying gossip stats. Unable to acquire lock.");
+            }
             epoch_index += 1;
         }
     }
 
-    async fn gossip_headers(&self, gossip_batch: &mut GossipBatch) -> anyhow::Result<()> {
-        info!("Gossiping headers in range: {:?}", gossip_batch.range);
-        if gossip_batch.range.end <= MERGE_BLOCK_NUMBER {
-            self.gossip_premerge_headers(gossip_batch).await
-        } else {
-            self.gossip_postmerge_headers(gossip_batch).await
+    async fn serve_full_block(
+        height: u64,
+        epoch_acc: Option<Arc<EpochAccumulator>>,
+        gossip_stats: Arc<Mutex<GossipStats>>,
+        portal_clients: Vec<HttpClient>,
+    ) -> anyhow::Result<()> {
+        debug!("Serving block: {height}");
+        let mut full_header = Bridge::get_header(height).await?;
+        if full_header.header.number <= MERGE_BLOCK_NUMBER {
```

Comment:
I think those kinds of comparisons reveal that we may need multiple `FullHeader`  versions represented as an enum.  I'm not saying that this is in the scope of this PR but something that we may think about or create an issue.

Reply:
Agree that your instinct is correct here. However, I'm not convinced that the enum overhead validates implementation for a fairly simple check that we do twice. But, I'd be curious to see what it looks like if somebody wants to give it a go. ... tracking in #754